fedimint-api-client/src/api/global_api/with_request_hook.rs

```diff
@@ -0,0 +1,92 @@
+use std::collections::BTreeSet;
+use std::sync::Arc;
+
+use fedimint_core::core::ModuleInstanceId;
+use fedimint_core::module::ApiRequestErased;
+use fedimint_core::task::{MaybeSend, MaybeSync};
+use fedimint_core::{apply, async_trait_maybe_send, maybe_add_send_sync, PeerId};
+use serde_json::Value;
+
+use super::super::{DynModuleApi, IRawFederationApi};
+use crate::api::PeerResult;
+
+/// "Api Request Hook"
+///
+/// An "api request hook" is a function that gets a raw federation api and
+/// can either pass it unmodified (no hook) or wrap it in whatever custom
+/// logic and return as a new raw federation api, possibly forwarding the call
+/// to the original one.
+///
+/// This is meant to allow downstream users to add custom logic for debugging,
+/// testing (e.g. simulating network being down), collecting stats, notifing
+/// about slow calls, errors, etc.
+pub type ApiRequestHook =
+    Arc<maybe_add_send_sync!(dyn Fn(DynIRawFederationApi) -> DynIRawFederationApi + 'static)>;
+
+pub type DynIRawFederationApi = Box<maybe_add_send_sync!(dyn IRawFederationApi + 'static)>;
+
+/// Convenience extension trait used for wrapping [`IRawFederationApi`] in
+/// a [`RawFederationApiWithRequestHook`]
+pub trait RawFederationApiWithRequestHookExt
+where
+    Self: Sized,
+{
+    fn with_request_hook(self, hook: &ApiRequestHook) -> RawFederationApiWithRequestHook;
+}
+
+impl<T> RawFederationApiWithRequestHookExt for T
+where
+    T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+{
+    fn with_request_hook(self, hook: &ApiRequestHook) -> RawFederationApiWithRequestHook {
+        RawFederationApiWithRequestHook::new(self, hook)
+    }
+}
+
+/// [`IRawFederationApi`] wrapping some `T: IRawFederationApi` and adding
+/// a toggle for online/offline state.
+///
+/// If the state is offline, all requests will block until toggle switched to
+/// online, without actually making any connections.
+///
+/// Use [`RawFederationApiWithRequestHookExt::with_request_hook`] to
+/// create.
+#[derive(Debug)]
+pub struct RawFederationApiWithRequestHook {
+    pub(crate) inner: DynIRawFederationApi,
+}
+
+impl RawFederationApiWithRequestHook {
+    pub fn new<T>(inner: T, hook: &ApiRequestHook) -> RawFederationApiWithRequestHook
+    where
+        T: IRawFederationApi + MaybeSend + MaybeSync + 'static,
+    {
+        RawFederationApiWithRequestHook {
+            inner: hook(Box::new(inner)),
+        }
+    }
+}
```

Comment:
Why even have a new struct if it only wraps a `DynIRawFederationApi`?

Reply:
Nicer API. The hook itself is just a function and can be passed around. It also can be carried over to a new client after restart, etc.